packages/flutter/lib/src/material/tabs.dart

```diff
@@ -432,63 +474,77 @@ class TabBarSelection {
     _performance
       ..progress = progress
       ..forward().then((_) {
-        if (onChanged != null)
-          onChanged();
+        if (config.onChanged != null)
+          config.onChanged(_index);
         _indexIsChanging = false;
       });
   }
 
   int get previousIndex => _previousIndex;
   int _previousIndex = 0;
+
+  Widget build(BuildContext context) {
+    return new _TabBarSelection(selection: this, child: config.child);
+  }
 }
 
-/// A tab strip, consisting of several TabLabels and a TabBarSelection.
-/// The TabBarSelection can be used to link this to a TabBarView.
+
+/// Displays a horizontal row of tabs, one per label. If isScrollable is
+/// true then each tab is as wide as needed for its label and the entire
+/// [TabBar] is scrollable. Otherwise each tab gets an equal share of the
+/// available space. A [TabBarSelection] widget ancestor must have been
+/// built to enable saving and monitoring the selected tab.
 ///
 /// Tabs must always have an ancestor Material object.
 class TabBar extends Scrollable {
   TabBar({
     Key key,
     this.labels,
-    this.selection,
     this.isScrollable: false
   }) : super(key: key, scrollDirection: ScrollDirection.horizontal) {
     assert(labels != null);
     assert(labels.length > 1);
-    assert(selection != null);
-    assert(selection.maxIndex == labels.length - 1);
   }
 
   final Iterable<TabLabel> labels;
-  final TabBarSelection selection;
   final bool isScrollable;
 
   _TabBarState createState() => new _TabBarState();
 }
 
 class _TabBarState extends ScrollableState<TabBar> {
+
+  TabBarSelectionState _selection;
+  bool _indexIsChanging = false;
+
+  int get _tabCount => config.labels.length;
+
+  void _addSelectionListeners(TabBarSelectionState selection) {
```

Comment:
Thinking about it more, I think the following test would leak:

Create a tab selection with a global key and a tab bar with a global key.
Have the tab selection be inserted by a StatefulBuilder whose child is some other object with a global key whose child is the tab bar.
Trigger a rebuild of the StatefulBuilder, and this time don't include the tab bar selection.
I don't think the tab bar view would be rebuilt, not would it find out about the change in the tab bar selection.
At this point I think you have a leak.
Insert the tab bar selection elsewhere in the tree, with another tab bar child.
Drive the tab bar selection.
I think both tab bars will actually move. Or at least, the first one will be rebuilt at least once.

I'm not sure what the right way to drive this is. If it wasn't a perf nightmare, I think the performance's progress should maybe be the state given to the _TabBarSelection inherited widget. But that would actually walk the whole tree each frame which seems crazy (though I guess it's what we're going to do with Theme animations, so...).

@abarth may have better ideas for how to handle this.

Maybe the tab bar selection can notify all descendants during its deactivate()?
Maybe we should subscribe to the TabBarSelection rather than performance?